### ISSUE TITLE
Beam Sync Support

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-client#136]
### Overview

Beam sync is an evolution of fast sync which starts sync on a launch block and only requesting state on demand and then slowly back-fill missing state. This allows to get to a head-of-chain execution within minutes while retaining healthy peer-to-peer networking properties to a greater extend.

An implementation in the client will allow for substantial new use cases in research, testing and dev tooling and brings the client closer to be run in a *stateless* execution context.

Current implementations of beam sync exist in [Trinity](https://github.com/ethereum/trinity) and the [Nethermind](https://github.com/NethermindEth/nethermind) client.

This is a tracking issue for the different tasks to be completed.

### Sources

- [Intro to Beam Sync - Jason Carver](https://medium.com/@jason.carver/intro-to-beam-sync-a0fd168be14a), medium Article, 2019-0924
- [Beam Sync Phase 0 (build upon fast sync) Specification](https://github.com/ethereum/stateless-ethereum-specs/blob/master/beam-sync-phase0.md), GitHub Stateless Ethereum Specs
- Ethereum Blog - ["beam sync"](https://blog.ethereum.org/search/?query=beam%20sync) Search
- Trinity - ["beam"](https://github.com/ethereum/trinity/search?q=beam&type=Issues)  GitHub Issue Search
- Nethermind - ["beam"](https://github.com/NethermindEth/nethermind/search?q=beam&type=Issues) GitHub Issue Search

### TODO

[ ] TODO
[ ] TODO

**Comment by s1na** [comment]
Being able to beam sync will be amazing. I do think however that a full sync should come before. It's the most simple kind of sync and the logic (block processing, re-org, etc.) is needed for all other syncs, because even fast sync and beam sync transition to full sync after they reach the tip. So just from a roadmap PoV I think it might make sense to target that first.

**Comment by holgerd77** [comment]
I would give this a yes and no.

My argumentation that it might make sense to start early on with a beam sync implementation (as well):

1. **Beam sync doesn't need to be fully implemented to be useful.** If we are on the level that state is downloaded along the recent tx/block needs we could already start integrating the VM into the client, which would (Attention! Highly subjective opinion! 😄 ) open up towards the most interesting use cases of the client like being able to real-world-test the VM internally or enable custom code execution and testing for smart contract developers on top of a recent network state. All this doesn't need the state backfill mechanism from beam sync or more complex mechanisms like pivoting and starting all over if state couldn't be retrieved within a certain range of blocks and it gets unlikely to catch up.

2. **Beam sync generalizes on existing sync concepts.** It e.g. needs a more flexible state downloader to be able to partially download state instead of a whole state snapshot. It might make sense to at least think beam sync along also during beam sync-independent tasks to have more generalized structures set up directly and prevent future double work, if e.g. a potential future too static state downloader needs a complete refactoring to then match the beam sync needs in the future.

That said: I think we very well should be able to find some common ground on the roadmap side. We can e.g. give the decision how far we go with a beam sync implementation some more time but already start thinking this along the implementations of the more general concepts. When we (likely: @cgewecke) work on the state downloader, we can e.g. already think about a more generalized architecture which would also work under beam sync conditions.

Just played around with the Nethermind client a bit and realized that we also likely need a block pivot (another pivot than the beam sync pivot here) mechanism, so to start (fast) syncing on a certain recent block with a hardcoded root hash as a validity proof (see [this](https://github.com/NethermindEth/nethermind/blob/1e97abc131583f07ab40478b138cabf9599499b3/src/Nethermind/Nethermind.Runner/configs/rinkeby.cfg#L29) nethermind config e.g.), otherwise also fast sync would take too much time (just did an estimate: ~5 hours for Rinkeby on the current client speed) to download all block headers.

These type of things - likely others as well - can be implemented already and being useful/necessary in a more general context.

We can give this some discussion in the next team call as well.

**Comment by s1na** [comment]
The (only? :D) good thing about full sync is that you don't need to download state. Integrate VM and start executing from block 0, .... I'm still of the opinion this is a hard requirement of beam sync. But that said I definitely agree with beam sync being the main vision.

**Comment by holgerd77** [comment]
@s1na mentioned in the chat from the call yesterday, that he would need the older HF implementations for his work. Then let's actually do this and start with implementing full sync here as Sina suggested. Implementing the old HF rules is on the table for quite some time anyhow - see https://github.com/ethereumjs/ethereumjs-vm/issues/652 - and apart from the VM integration (needed anyhow) full sync addition shouldn't cause a lot of overhead.

Since state download is delayed a bit there is also additional reasoning here from a pm perspective: then we can start earlier on the VM integration. Aaaaand: I do have to admit that it makes a lot of sense from a VM testing perspective to just run the VM as long as possible, even if full sync might not have too much relevant production use. Since this is the ultimate test bed to just run ALL the block ever occurred on the network.

I'll directly open a `SpuriousDragon` branch on this on the VM and give this a start. This might be a more-talk-than-work topic anyhow (but might very well also be the other way around, let's see 😛 ).

**Comment by jochem-brouwer** [comment]
I agree that Full Sync makes sense to implement first, but note that if we use Beam Sync on the current mainnet we can essentially use the VM out of the box (since we do not have to execute currently "unsupported" blocks (i.e. before Byzantium)). 

I think I made this point on the call too: Beam Sync picks a "tip" / pivot point of the chain and starts executing this block. If it "lags behind too far" it will then pause the VM and choose a new tip. I think we should export a method which forces the block to be executed using Beam Sync. 

Why? If you create a transaction on any network and want to debug said transaction, you can then use the "beam sync" method to re-run said block (and transactions), which includes proofs of everything. This is better than using `web3` to get the storage slot values / account balances, etc. as this does not include a proof. You can thus be reasonably certain that the block you are then running is the "correct" one. 

This has massive value. Imagine that you don't have to run a node, use the Light Ethereum Protocol to quickly get the necessary data to run a block, and then being able to debug said transaction. It can be even be done from a browser. Maybe it is not clear from reading this text, but I can barely contain my excitement about this 😄 .

Point to take from here: export the beam sync "run block x" function so that users can actually force the client to quickly "beam-run" (we need a better term for this I guess) a block/transaction.

I'll spend the next weeks a bit to get updated on this amazing package and I'll start spending some time thinking about implementing Beam Sync.

**Comment by jochem-brouwer** [comment]
I will dump here my thoughts on how beam sync should be implemented.

- Beam sync needs to have a specific class in which handles running a block in beam sync. That way we can also run the client and tell it, "run block X".
  - It also needs a way to stop running the current block, e.g. when we are actually beam syncing and the current block becomes stale (we switch the pivot block)
  - The BeamSync should be able to run both FastSync and BeamSync in parallel (but these can thus be developed independently)
- To run BeamSync we can use a specific `StateManager`. I'd say we import the `StateManager` from VM and create a class inside client; `BeamStateManager extends StateManager` where we can just run the StateManager by default but we overload the functionality (the `get` operations) where if a key is not found, we request this data over ETH protocol:
- Client needs ETH protocol support for `GetNodeData` and `NodeData`
- MPT needs a feature where it reports missing nodes. Right now, if a key is not found in the DB, the walk controller throws. We want an option where we can asynchronously try to get a key: if a key is not found, this would call some function (`OnNodeMissing`) where we can then request this data via `GetNodeData`, put it in the DB and then call back on the walk controller such that we can get the next key (or, if we are done, we just found the value, and give control back to `StateManager`)
